src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -501,6 +502,17 @@ void MergeTreeData::checkProperties(
 
         for (const auto & index : new_metadata.secondary_indices)
         {
+            if (!allow_suspicious_indices && !attach)
+            {
+                NameSet index_key_columns_set;
+                const auto & column_names = index.column_names;
+
+                for (size_t i = 0; i < column_names.size(); ++i)
+                {
+                    if(!index_key_columns_set.emplace(column_names[i]).second)
+                        throw Exception("Secondary indices contains duplicate columns", ErrorCodes::BAD_ARGUMENTS);
+                }
+             }
```

Comment:
Maybe it makes sense to compare definitions themselves?
You can use something like
```
auto hash = ast_definition.getTreeHash();
String key = toString(hash.first) + '_' + toString(hash.second);
```

Reply:
Updated